### PR TITLE
[inductor] Misc refactors

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126417
* #127047
* __->__ #126945
* #126944

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by shunting314 — APPROVED** [review]

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -536,6 +537,28 @@ def reciprocal(x):
...
+    def expm1(x):
+        return ops.sub(ops.exp(x), ops.constant(1, torch.float32))
+
+    @staticmethod
+    def log10(x):
+        return ops.mul(ops.log(x), ops.constant(1 / math.log(10), torch.float32))
+
+    @staticmethod
+    def log2(x):
+        return ops.mul(ops.log(x), ops.constant(1 / math.log(2), torch.float32))
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/codegen/common.py:558]: >  If some new backend doesn't have a native log2 function they can use this fallback.

Yes this is the issue I'm concerned about. The new backend will seem to work fine but silently have precision issues with these functions.

File: torch/_inductor/codegen/simd.py [hunk 2]
```diff
@@ -851,8 +851,9 @@ def mask_loads(self, mask):
         """Context manager to add an additional mask to tl.load/store"""
         prior = self._load_mask
         if prior:
-            mask = self.cse.generate(self.compute, f"{mask} & {prior}")
+            mask = ops.logical_and(mask, prior)
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/codegen/simd.py:854]: Correct me if I'm wrong. CSE generation need take a buffer to store the new expression and it returns the variable name for the expression. `ops.*` does not takes a buffer, so I assume it returns the expression itself no matter if we have seen that expression before.

[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/codegen/simd.py:854]: `ops.` calls the current ops handler, which is `CSEProxy` in `codegen/common.py` which does CSE on the compute buffer by default

[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/codegen/simd.py:854]: oh, I see. We implicitly use compute buffer by default. Make sense now.

File: torch/_inductor/codegen/common.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/codegen/common.py:558]: Is there an alternative for backends that don't have a native log2?

**Pull request closed without merging**